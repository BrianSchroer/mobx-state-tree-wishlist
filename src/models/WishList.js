import { types } from 'mobx-state-tree';

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: '' // optional string - default to '' - shorthand for "image: types.optional(types.string, '')"
  }
  ).actions(self => ({
    changeName: newName => self.name = newName,
    changePrice: newPrice => self.price = newPrice,
    changeImage: newImage => self.image = newImage
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  }
  ).actions(self => ({
    add: item => self.items.push(item)
  }));