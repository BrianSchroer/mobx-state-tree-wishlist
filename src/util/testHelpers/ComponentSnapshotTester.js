import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component snapshot helper
 */
export default class ComponentSnapshotTester {

  /**
   * Create new ComponentSnapshotTester instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propOverrides) => Object.assign({}, props, propOverrides);
  }

  /**
   * Specify callback function to be called by the test function to adjust component
   * props with propOverrides.
   * (Defaults to adjusting top-level props.)
   */
  withPropAdjustor = (adjustor) => {
    this.adjustProps = adjustor;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propOverrides, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = (propOverrides) => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propOverrides);

    const newSnapshot = renderer.create(<Component {...props} />).toJSON();

    expect(newSnapshot).toMatchSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot. 
   * @param {*} component - React component1 
   */
  static test(component) {
    new ComponentSnapshotTester(component).test();
  }
}