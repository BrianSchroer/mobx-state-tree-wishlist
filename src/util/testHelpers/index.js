export { default as mobxSnapshotHelper } from './mobxSnapshotHelper/snapshotHelper';
export { default as ComponentSnapshotHelper } from './ComponentSnapshotHelper/SnapshotHelper';