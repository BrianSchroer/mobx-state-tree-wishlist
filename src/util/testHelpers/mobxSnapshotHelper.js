import { getSnapshot } from 'mobx-state-tree';

export default class {
  static createSnapshot(target) {
    return getSnapshot(target);
  }

  static assertMatch(target) {
    expect(this.createSnapshot(target)).toMatchSnapshot();
  }
}