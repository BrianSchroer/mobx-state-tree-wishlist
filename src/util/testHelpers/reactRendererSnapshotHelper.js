import renderer from 'react-test-renderer';

/**
 * react-test-renderer snapshot helper methods
 */
export default class {

  /**
   * Create component from JSX
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {object} component object
   */
  static createComponent(componentJsx) {
    return renderer.create(componentJsx);
  }

  /**
   * Create component from JSX and serialize as JSON "snapshot" string.
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {string} JSON string.
   */
  static createSnapshot(componentJsx) {
    return this.createComponent(componentJsx).toJSON();
  }

  /**
   * Create component from JSX, serialize as JSON string, and compare to accepted snapshot.
   * If a snapshot doesn't already exist, one is saved. If a previous snapshot exists
   * and the new snapshot doesn't match, you'll be asked if that's an error or if you want
   * to accept the new snapshot.
   * @param {JSX} componentJsx (e.g. '<Button />')
   */
  static assertMatch(componentJsx) {
    expect(this.createSnapshot(componentJsx)).toMatchSnapshot();
  }
}