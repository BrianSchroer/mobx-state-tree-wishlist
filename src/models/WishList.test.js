import { WishList, WishListItem } from './WishList';
import { mobxSnapshotHelper as snapshotHelper } from '../util/testHelpers';

const testInput = {
  name: 'Chronicles of Narnia Box Set - C.S. Lewis',
  price: 28.83,
  image: 'https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg'
};

describe('WishListItem', () => {
  it('.create() should return expected values', () => {
    snapshotHelper.assertMatch(WishListItem.create(testInput));
  });

  it('.create() without image should default to empty string for image', () => {
    const input = {
      name: 'Chronicles of Narnia Box Set - C.S. Lewis',
      price: 28.83
    };

    snapshotHelper.assertMatch(WishListItem.create(input));
  });

  it('.changeName() should change name', () => {
    const output = WishListItem.create(testInput);
    output.changeName('new name');
    expect(output.name).toBe('new name');
  });

  it('.changePrice() should change price', () => {
    const output = WishListItem.create(testInput);
    output.changePrice(12.34);
    expect(output.price).toBe(12.34);
  });

  it('.changeImage() should change image', () => {
    const output = WishListItem.create(testInput);
    output.changeImage('new image');
    expect(output.image).toBe('new image');
  });
});

describe('WishList', () => {
  it('.create() should returns expected items', () => {
    const inputItems = [1, 2, 3].map(i => ({ name: `name${i}`, price: i, image: `image${i}` }));
    snapshotHelper.assertMatch(WishList.create({ items: inputItems }).items);
  });

  it('.create() without input returns empty items array', () => {
    const output = WishList.create();
    expect(output.items.length).toBe(0);
  });

  it('.add(object) should add item', () => {
    const wishList = WishList.create();
    const countBeforeAdd = wishList.items.length;
    const newItem = { name: 'newName', price: 23.45, image: 'new image' };
    wishList.add(newItem);
    expect(wishList.items.length).toEqual(countBeforeAdd + 1);
    snapshotHelper.assertMatch(wishList.items[countBeforeAdd]);
  });

  it('.add(WishListItem) should add item', () => {
    const wishList = WishList.create();
    const countBeforeAdd = wishList.items.length;
    const newItem = WishListItem.create({ name: 'newName', price: 23.45, image: 'new image' });
    wishList.add(newItem);
    expect(wishList.items.length).toEqual(countBeforeAdd + 1);
    snapshotHelper.assertMatch(wishList.items[countBeforeAdd]);
  });

});