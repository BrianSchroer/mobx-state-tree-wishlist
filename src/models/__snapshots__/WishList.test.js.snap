// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishList .add(WishListItem) should add item 1`] = `
Object {
  "image": "new image",
  "name": "newName",
  "price": 23.45,
}
`;

exports[`WishList .add(object) should add item 1`] = `
Object {
  "image": "new image",
  "name": "newName",
  "price": 23.45,
}
`;

exports[`WishList .create() should returns expected items 1`] = `
Array [
  Object {
    "image": "image1",
    "name": "name1",
    "price": 1,
  },
  Object {
    "image": "image2",
    "name": "name2",
    "price": 2,
  },
  Object {
    "image": "image3",
    "name": "name3",
    "price": 3,
  },
]
`;

exports[`WishListItem .create() should return expected values 1`] = `
Object {
  "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
  "name": "Chronicles of Narnia Box Set - C.S. Lewis",
  "price": 28.83,
}
`;

exports[`WishListItem .create() without image should default to empty string for image 1`] = `
Object {
  "image": "",
  "name": "Chronicles of Narnia Box Set - C.S. Lewis",
  "price": 28.83,
}
`;
